In [1]:
#Libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.preprocessing import OneHotEncoder
from sklearn.preprocessing import OrdinalEncoder

In [2]:
df = pd.read_csv(r"C:\Users\EliteBook\OneDrive\Desktop\Sem III\ML - Lab\A2\student_depression_dataset.csv")
df.head()

id  Gender   Age           City Profession  Academic Pressure  \
0   2    Male  33.0  Visakhapatnam    Student                5.0   
1   8  Female  24.0      Bangalore    Student                2.0   
2  26    Male  31.0       Srinagar    Student                3.0   
3  30  Female  28.0       Varanasi    Student                3.0   
4  32  Female  25.0         Jaipur    Student                4.0   

   Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0            0.0  8.97                 2.0               0.0   
1            0.0  5.90                 5.0               0.0   
2            0.0  7.03                 5.0               0.0   
3            0.0  5.59                 2.0               0.0   
4            0.0  8.13                 3.0               0.0   

        Sleep Duration Dietary Habits   Degree  \
0          '5-6 hours'        Healthy  B.Pharm   
1          '5-6 hours'       Moderate      BSc   
2  'Less than 5 hours'        Healthy       BA   
3          '7-8 hours'       Moderate      BCA   
4          '5-6 hours'       Moderate   M.Tech   

  Have you ever had suicidal thoughts ?  Work/Study Hours Financial Stress  \
0                                   Yes               3.0              1.0   
1                                    No               3.0              2.0   
2                                    No               9.0              1.0   
3                                   Yes               4.0              5.0   
4                                   Yes               1.0              1.0   

  Family History of Mental Illness  Depression  
0                               No           1  
1                              Yes           0  
2                              Yes           0  
3                              Yes           1  
4                               No           0

In [3]:
df['Depression'].value_counts()

Depression
1    16336
0    11565
Name: count, dtype: int64

In [4]:
# The dataset exhibits a mild class imbalance.

In [5]:
df['Financial Stress'] = df['Financial Stress'].replace('?', np.nan)
df['Financial Stress'] = pd.to_numeric(df['Financial Stress'])
df.dropna(inplace=True)

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

In [7]:
drop = ['id','City', 'Profession','Degree']
df1 =df.drop(columns = drop)
df1

Gender   Age  Academic Pressure  Work Pressure  CGPA  \
0        Male  33.0                5.0            0.0  8.97   
1      Female  24.0                2.0            0.0  5.90   
2        Male  31.0                3.0            0.0  7.03   
3      Female  28.0                3.0            0.0  5.59   
4      Female  25.0                4.0            0.0  8.13   
...       ...   ...                ...            ...   ...   
27896  Female  27.0                5.0            0.0  5.75   
27897    Male  27.0                2.0            0.0  9.40   
27898    Male  31.0                3.0            0.0  6.61   
27899  Female  18.0                5.0            0.0  6.88   
27900    Male  27.0                4.0            0.0  9.24   

       Study Satisfaction  Job Satisfaction       Sleep Duration  \
0                     2.0               0.0          '5-6 hours'   
1                     5.0               0.0          '5-6 hours'   
2                     5.0               0.0  'Less than 5 hours'   
3                     2.0               0.0          '7-8 hours'   
4                     3.0               0.0          '5-6 hours'   
...                   ...               ...                  ...   
27896                 5.0               0.0          '5-6 hours'   
27897                 3.0               0.0  'Less than 5 hours'   
27898                 4.0               0.0          '5-6 hours'   
27899                 2.0               0.0  'Less than 5 hours'   
27900                 1.0               0.0  'Less than 5 hours'   

      Dietary Habits Have you ever had suicidal thoughts ?  Work/Study Hours  \
0            Healthy                                   Yes               3.0   
1           Moderate                                    No               3.0   
2            Healthy                                    No               9.0   
3           Moderate                                   Yes               4.0   
4           Moderate                                   Yes               1.0   
...              ...                                   ...               ...   
27896      Unhealthy                                   Yes               7.0   
27897        Healthy                                    No               0.0   
27898      Unhealthy                                    No              12.0   
27899        Healthy                                   Yes              10.0   
27900        Healthy                                   Yes               2.0   

       Financial Stress Family History of Mental Illness  Depression  
0                   1.0                               No           1  
1                   2.0                              Yes           0  
2                   1.0                              Yes           0  
3                   5.0                              Yes           1  
4                   1.0                               No           0  
...                 ...                              ...         ...  
27896               1.0                              Yes           0  
27897               3.0                              Yes           0  
27898               2.0                               No           0  
27899               5.0                               No           1  
27900               3.0                              Yes           1  

[27898 rows x 14 columns]

In [8]:
enc = OrdinalEncoder()
cat1 = ['Have you ever had suicidal thoughts ?','Family History of Mental Illness','Gender']
df1[cat1]= enc.fit_transform(df1[cat1]).astype('int')
df1
cat2 = ['Dietary Habits','Sleep Duration']
dummies = pd.get_dummies(df1[cat2],drop_first= True).astype(int)
df2 = df1.drop(columns = cat2).join(dummies)
df2.head(6)

Gender   Age  Academic Pressure  Work Pressure  CGPA  Study Satisfaction  \
0       1  33.0                5.0            0.0  8.97                 2.0   
1       0  24.0                2.0            0.0  5.90                 5.0   
2       1  31.0                3.0            0.0  7.03                 5.0   
3       0  28.0                3.0            0.0  5.59                 2.0   
4       0  25.0                4.0            0.0  8.13                 3.0   
5       1  29.0                2.0            0.0  5.70                 3.0   

   Job Satisfaction  Have you ever had suicidal thoughts ?  Work/Study Hours  \
0               0.0                                      1               3.0   
1               0.0                                      0               3.0   
2               0.0                                      0               9.0   
3               0.0                                      1               4.0   
4               0.0                                      1               1.0   
5               0.0                                      0               4.0   

   Financial Stress  Family History of Mental Illness  Depression  \
0               1.0                                 0           1   
1               2.0                                 1           0   
2               1.0                                 1           0   
3               5.0                                 1           1   
4               1.0                                 0           0   
5               1.0                                 0           0   

   Dietary Habits_Moderate  Dietary Habits_Others  Dietary Habits_Unhealthy  \
0                        0                      0                         0   
1                        1                      0                         0   
2                        0                      0                         0   
3                        1                      0                         0   
4                        1                      0                         0   
5                        0                      0                         0   

   Sleep Duration_'7-8 hours'  Sleep Duration_'Less than 5 hours'  \
0                           0                                   0   
1                           0                                   0   
2                           0                                   1   
3                           1                                   0   
4                           0                                   0   
5                           0                                   1   

   Sleep Duration_'More than 8 hours'  Sleep Duration_Others  
0                                   0                      0  
1                                   0                      0  
2                                   0                      0  
3                                   0                      0  
4                                   0                      0  
5                                   0                      0

In [9]:
x = df2.drop('Depression', axis=1)
y = df2['Depression']
rand = 42
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size = 0.2,random_state = rand)
np.round(len(x_train)/len(x),3)

0.8

In [15]:
scaler = StandardScaler()

categorics = ['Gender', 'Have you ever had suicidal thoughts ?','Family History of Mental Illness',
       'Dietary Habits_Moderate', 'Dietary Habits_Others',
       'Dietary Habits_Unhealthy',"Sleep Duration_'7-8 hours'",
       "Sleep Duration_'Less than 5 hours'",
       "Sleep Duration_'More than 8 hours'", 'Sleep Duration_Others']
x_train_scaled = scaler.fit_transform(x_train.drop(categorics,axis=1))
x_test_scaled = scaler.transform(x_test.drop(categorics, axis=1))

In [21]:
x_train_scaled = pd.DataFrame(x_train_scaled, columns=x_train.drop(categorics,axis=1).columns)
x_train_scaled[categorics] = x_train[categorics].reset_index(drop = True)
x_train_scaled
x_test_scaled = pd.DataFrame(x_test_scaled,columns=x_test.drop(categorics,axis=1).columns)
x_test_scaled[categorics] = x_test[categorics].reset_index(drop=True)
x_test_scaled

Age  Academic Pressure  Work Pressure      CGPA  \
0     0.240645          -0.821130      -0.009467  0.237568   
1    -1.394588           1.349595      -0.009467 -0.790701   
2    -1.394588          -0.097555      -0.009467 -1.308241   
3     0.036241           1.349595      -0.009467 -1.703205   
4    -0.372567          -0.821130      -0.009467  0.530386   
...        ...                ...            ...       ...   
5575  1.467070           1.349595      -0.009467 -0.600029   
5576 -0.576971          -1.544706      -0.009467  0.632532   
5577  1.058262          -0.821130      -0.009467  1.224979   
5578  0.445049           0.626020      -0.009467 -0.286781   
5579 -0.372567           1.349595      -0.009467  0.809585   

      Study Satisfaction  Job Satisfaction  Work/Study Hours  \
0               1.515229         -0.016399         -0.577803   
1               0.043426         -0.016399          0.498810   
2               0.043426         -0.016399         -1.385263   
3              -1.428377         -0.016399          1.306271   
4               0.043426         -0.016399          1.037117   
...                  ...               ...               ...   
5575           -1.428377         -0.016399          0.229657   
5576            0.779327         -0.016399          1.037117   
5577            1.515229         -0.016399          1.037117   
5578           -1.428377         -0.016399         -1.385263   
5579            0.043426         -0.016399         -0.039496   

      Financial Stress  Gender  Have you ever had suicidal thoughts ?  \
0             1.297106       0                                      1   
1            -1.486166       1                                      1   
2            -0.094530       1                                      1   
3             1.297106       0                                      1   
4            -1.486166       1                                      1   
...                ...     ...                                    ...   
5575         -1.486166       1                                      1   
5576         -0.094530       1                                      0   
5577          0.601288       1                                      0   
5578         -0.094530       1                                      1   
5579         -0.790348       1                                      1   

      Family History of Mental Illness  Dietary Habits_Moderate  \
0                                    0                        0   
1                                    0                        0   
2                                    0                        0   
3                                    0                        0   
4                                    1                        0   
...                                ...                      ...   
5575                                 1                        1   
5576                                 0                        0   
5577                                 1                        0   
5578                                 1                        0   
5579                                 0                        1   

      Dietary Habits_Others  Dietary Habits_Unhealthy  \
0                         0                         1   
1                         0                         1   
2                         0                         1   
3                         0                         1   
4                         0                         1   
...                     ...                       ...   
5575                      0                         0   
5576                      0                         0   
5577                      0                         1   
5578                      0                         1   
5579                      0                         0   

      Sleep Duration_'7-8 hours'  Sleep Duration_'Less than 5 hours'  \
0                              1                          

In [23]:
## Logistic Regression

In [25]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

logistic = LogisticRegression()
log = logistic.fit(x_train_scaled, y_train)
log

LogisticRegression()

In [26]:
y_proba = logistic.predict_proba(x_test_scaled)[:, 1]

# Evaluate with different thresholds
for prob in [0.3, 0.5, 0.7]:
    y_pred_prob = (y_proba >= prob).astype(int)
    
    print(f"Threshold = {prob}")
    print("Accuracy:", accuracy_score(y_test, y_pred_prob))
    print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_prob))

Threshold = 0.3
Accuracy: 0.8290322580645161
Confusion Matrix:
 [[1536  812]
 [ 142 3090]]
Threshold = 0.5
Accuracy: 0.8448028673835125
Confusion Matrix:
 [[1823  525]
 [ 341 2891]]
Threshold = 0.7
Accuracy: 0.8295698924731183
Confusion Matrix:
 [[2047  301]
 [ 650 2582]]


#### Interpretation:

Varying the classification threshold impacts the trade-off between precision and recall.

Threshold 0.3: High recall but many false positives — good for catching most depression cases, but with more misclassifications.

Threshold 0.5: Offers the best balance between precision and recall with the highest accuracy (78.51%).

Threshold 0.7: Higher precision, fewer false positives, but misses many actual cases (low recall).

##### Conclusion:

Threshold 0.5 is the most balanced choice, but the optimal threshold should align with the application's priority — minimizing false negatives or false positives.

In [30]:
y_pred_thresh = (y_proba >= 0.3).astype(int)
y_pred_thresh.sum()

3902

In [32]:
y_pred_thresh = (y_proba >= 0.5).astype(int)
y_pred_thresh.sum()

3416

In [34]:
y_pred_thresh = (y_proba >= 0.7).astype(int)
y_pred_thresh.sum()

2883

In [36]:
(y_pred_thresh == 0).sum()  # Non-depressed 


2697

In [38]:
y_hat_30 = (y_proba >= 0.3)

In [42]:
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score, matthews_corrcoef
from sklearn.metrics import confusion_matrix, classification_report, roc_curve, roc_auc_score

In [44]:
print(classification_report(y_test, y_hat_30))

              precision    recall  f1-score   support

           0       0.92      0.65      0.76      2348
           1       0.79      0.96      0.87      3232

    accuracy                           0.83      5580
   macro avg       0.85      0.81      0.81      5580
weighted avg       0.84      0.83      0.82      5580



In [180]:
## KNN Regression

In [156]:
from sklearn.neighbors import KNeighborsClassifier
KNN_classifier = KNeighborsClassifier(n_neighbors=k)
KNN_classifier.fit(x_train, y_train)

KNeighborsClassifier(n_neighbors=20)

In [158]:
y_hat      = KNN_classifier.predict(x_test)
y_hat_probs = KNN_classifier.predict_proba(x_test)[:,1]

In [159]:
np.max(y_hat_probs)

1.0

In [160]:
y_hat_30 = np.where(y_hat_probs>0.3,1,0)
y_hat_50 = np.where(y_hat_probs>0.5,1,0)
y_hat_70 = np.where(y_hat_probs>0.7,1,0)

In [161]:
df_predictions = pd.DataFrame({'y_test':y_test, 'y_hat_probs':y_hat_probs,
                               'y_hat_30':y_hat_30, 'y_hat_50':y_hat_50,'y_hat_70':y_hat_70})
df_predictions.head()

y_test  y_hat_probs  y_hat_30  y_hat_50  y_hat_70
16479       1         0.80         1         1         1
19518       0         0.85         1         1         1
27260       1         0.80         1         1         1
8318        1         1.00         1         1         1
9889        1         0.30         0         0         0

In [162]:
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score
from sklearn.metrics import confusion_matrix, classification_report, roc_curve, roc_auc_score

In [163]:
print(confusion_matrix(y_test, y_hat))

[[1724  624]
 [ 420 2812]]


##### Results of Confusion Matrix

True Negatives (TN) = 1724: Correctly predicted non-depressed.

False Positives (FP) = 624: Incorrectly predicted as depressed (actually non-depressed).

False Negatives (FN) = 420: Incorrectly predicted as non-depressed (actually depressed).

True Positives (TP) = 2812: Correctly predicted depressed.

In [165]:
print(classification_report(y_test, y_hat))

              precision    recall  f1-score   support

           0       0.80      0.73      0.77      2348
           1       0.82      0.87      0.84      3232

    accuracy                           0.81      5580
   macro avg       0.81      0.80      0.81      5580
weighted avg       0.81      0.81      0.81      5580



#### Interpretation of Classification Report

This classification report assesses the performance of a binary model predicting depression status (class 1 for depressed, class 0 for not depressed) across 5,580 observations. The model achieved an overall accuracy of 81%, indicating that it correctly classified the depression status in the majority of cases, reflecting a strong understanding of the data.

For class 0 (No Depression), the model recorded a precision of 0.80 and a recall of 0.73. This suggests that 80% of the individuals predicted as non-depressed were indeed not depressed, though the model identified only 73% of all actual non-depressed individuals. The F1-score of 0.77 indicates reasonable performance, with slightly lower sensitivity in detecting non-depressed cases.

In contrast, the model performed better on class 1 (Depression), with a precision of 0.82 and a recall of 0.87. It successfully identified 87% of all actual depression cases, while maintaining strong accuracy in its positive predictions. The F1-score of 0.84 demonstrates a well-balanced and effective performance in detecting depression.

The macro averages, which treat both classes equally, reflect the slight drop in recall due to weaker performance on class 0. However, the weighted averages, which take class distribution into account, remain consistently high at 0.81, indicating the model's overall robustness even in the presence of class imbalance.

##### Conclusion

In conclusion, the model excels at detecting depression cases, with slightly reduced effectiveness in identifying non-depressed individuals a trade-off that may be acceptable in contexts where minimizing missed depression diagnoses is a priority.